### PR TITLE
feat: postgres lock

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Lock for DB migrations (or any other async operation) to run only from one instance at a time. The lock works for lock id + database name so will work with infra level multi-tenancy. 

`pg_advisory_lock` is like zookeeper inside postgres :) 

## Discussion points

* This lock has a timeout of 5 seconds. After 5 seconds of waiting for a lock no migration is run but the application startup proceeds. Another option would be to run a migration after 5 seconds anyway.
* It is behind a feature flag for now to keep it safe

### CODE REVIEW

**Review by sjaanus — APPROVED** [review]
Nice!

File: src/lib/util/db-lock.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+export const withDbLock =
+    (dbConfig: IDBOption, timeout = 5000, lockKey = defaultLockKey) =>
+    <A extends any[], R>(fn: (...args: A) => Promise<R>) =>
+    async (...args: A): Promise<R> => {
+        const client = new Client({ ...dbConfig, query_timeout: timeout });
+        try {
+            await client.connect();
+            // wait to obtain a lock
+            await client.query('SELECT pg_advisory_lock($1)', [lockKey]);
```
Comment (kwasniew) [inline src/lib/util/db-lock.ts:14]: only one instance with a given key and DB name can run the SELECT. Other clients will await on the lock to free up. The timeout of 5000 prevents from infinite waiting. Realistically our migrations should run in under 5 seconds.

File: src/lib/util/db-lock.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+            const result = await fn(...args);
+            return result;
+        } catch (e) {
+            console.error('Locking error:', e);
+            throw e;
+        } finally {
+            await client.query('SELECT pg_advisory_unlock($1)', [lockKey]);
```
Comment (kwasniew) [inline src/lib/util/db-lock.ts:21]: free up the lock

File: src/lib/util/db-lock.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+export const withDbLock =
+    (dbConfig: IDBOption, timeout = 5000, lockKey = defaultLockKey) =>
+    <A extends any[], R>(fn: (...args: A) => Promise<R>) =>
+    async (...args: A): Promise<R> => {
+        const client = new Client({ ...dbConfig, query_timeout: timeout });
+        try {
+            await client.connect();
+            // wait to obtain a lock
+            await client.query('SELECT pg_advisory_lock($1)', [lockKey]);
+            const result = await fn(...args);
```
Comment (kwasniew) [inline src/lib/util/db-lock.ts:15]: the original async function that we're protecting with a lock

**Pull request merged**